aws/src/main/java/org/apache/iceberg/aws/s3/DefaultS3FileIOAwsClientFactory.java

```diff
@@ -54,6 +56,11 @@ public S3Client s3() {
                     awsClientProperties, s3ClientBuilder))
         .applyMutation(s3FileIOProperties::applySignerConfiguration)
         .applyMutation(s3FileIOProperties::applyS3AccessGrantsConfigurations)
+        .overrideConfiguration(
+            config ->
+                config.putAdvancedOption(
+                    SdkAdvancedClientOption.USER_AGENT_PREFIX,
+                    "s3fileio " + IcebergBuild.fullVersion()))
```

Comment:
you might also need to update https://github.com/apache/iceberg/blob/556b79893c4cf760f031258f0d8ae657e42b4443/aws/src/main/java/org/apache/iceberg/aws/AwsClientFactories.java#L116

Reply:
also maybe worth changing the user agent to `"s3fileio/" + EnvironmentContext.get()`. That way if you're running Spark/Flink, you'll also get information about that engine and its version

Reply:
Thanks, these are great callouts! 

I factored this lambda out into a separate "mutator" like it is done for the other configs and included a unit test that at least calls the code. I tried to introduce some assertions to check if the correct String is set as User Agent but it felt like I was just re-implementing the original code path with captors so I abandoned that effort.